File: aeropinn/geometry/naca.py
"""NACA 4-digit airfoil geometry: thickness/camber distributions and surface coordinates.

Standard closed-form NACA 4-digit parameterization (open trailing edge).
m = max camber (fraction of chord), p = location of max camber (fraction of chord),
t = max thickness (fraction of chord). Chord length c = 1 (nondimensional).
"""
import numpy as np

_THICKNESS_COEFFS = (0.2969, -0.1260, -0.3516, 0.2843, -0.1015)


def naca_thickness(x: np.ndarray, t: float) -> np.ndarray:
    """Half-thickness distribution y_t(x) for 0 <= x <= 1."""
    x = np.asarray(x, dtype=np.float64)
    a0, a1, a2, a3, a4 = _THICKNESS_COEFFS
    return 5.0 * t * (a0 * np.sqrt(x) + a1 * x + a2 * x**2 + a3 * x**3 + a4 * x**4)


def naca_camber(x: np.ndarray, m: float, p: float) -> np.ndarray:
    """Camber line y_c(x). Returns zeros identically if m == 0 (symmetric airfoil)."""
    x = np.asarray(x, dtype=np.float64)
    if m == 0.0 or p == 0.0:
        return np.zeros_like(x)
    front = x < p
    yc = np.empty_like(x)
    yc[front] = (m / p**2) * (2 * p * x[front] - x[front] ** 2)
    yc[~front] = (m / (1 - p) ** 2) * ((1 - 2 * p) + 2 * p * x[~front] - x[~front] ** 2)
    return yc


def naca_camber_slope(x: np.ndarray, m: float, p: float) -> np.ndarray:
    """dy_c/dx, used to rotate thickness normal to the camber line."""
    x = np.asarray(x, dtype=np.float64)
    if m == 0.0 or p == 0.0:
        return np.zeros_like(x)
    front = x < p
    dyc = np.empty_like(x)
    dyc[front] = (2 * m / p**2) * (p - x[front])
    dyc[~front] = (2 * m / (1 - p) ** 2) * (p - x[~front])
    return dyc


def _cosine_spacing(n_points: int) -> np.ndarray:
    """Cosine-spaced x in [0, 1], denser near leading and trailing edge."""
    beta = np.linspace(0.0, np.pi, n_points)
    return 0.5 * (1.0 - np.cos(beta))


def naca4_surface(m: float, p: float, t: float, n_points: int = 200):
    """Upper and lower surface coordinates for a NACA 4-digit airfoil.

    Returns (upper_xy, lower_xy), each an (n_points, 2) array, ordered leading-edge
    to trailing-edge, cosine-spaced.
    """
    x = _cosine_spacing(n_points)
    yt = naca_thickness(x, t)
    yc = naca_camber(x, m, p)
    dyc = naca_camber_slope(x, m, p)
    theta = np.arctan(dyc)

    xu = x - yt * np.sin(theta)
    yu = yc + yt * np.cos(theta)
    xl = x + yt * np.sin(theta)
    yl = yc - yt * np.cos(theta)

    upper = np.stack([xu, yu], axis=-1)
    lower = np.stack([xl, yl], axis=-1)
    return upper, lower


def naca4_boundary_loop(m: float, p: float, t: float, n_points: int = 200) -> np.ndarray:
    """Closed boundary polygon: lower surface (LE->TE) then upper surface reversed (TE->LE).

    Returns an (2*n_points - 1, 2) array with no duplicated leading-edge point across
    the join (both branches start at the same LE point x=0).
    """
    upper, lower = naca4_surface(m, p, t, n_points)
    return np.concatenate([lower, upper[::-1][1:]], axis=0)

File: web/lib/naca.ts
/**
 * NACA 4-digit airfoil geometry — direct TypeScript port of aeropinn/geometry/naca.py.
 * Pure math, no dependencies. Used wherever the site renders a real airfoil outline
 * (Figure 01, Section 02's scroll scene, Section 03's precomputed-case views) so
 * geometry is always the same real parameterization the trained model was built on,
 * never a generic/stock airfoil shape.
 *
 * Verify against the Python source before trusting in a figure: sample a few (m,p,t,x)
 * points and diff naca_thickness/naca_camber against aeropinn/geometry/naca.py's output.
 */

export interface Point {
  x: number;
  y: number;
}

const THICKNESS_COEFFS = [0.2969, -0.126, -0.3516, 0.2843, -0.1015] as const;

export function nacaThickness(x: number, t: number): number {
  const [a0, a1, a2, a3, a4] = THICKNESS_COEFFS;
  return 5 * t * (a0 * Math.sqrt(x) + a1 * x + a2 * x ** 2 + a3 * x ** 3 + a4 * x ** 4);
}

export function nacaCamber(x: number, m: number, p: number): number {
  if (m === 0 || p === 0) return 0;
  if (x < p) {
    return (m / p ** 2) * (2 * p * x - x ** 2);
  }
  return (m / (1 - p) ** 2) * (1 - 2 * p + 2 * p * x - x ** 2);
}

export function nacaCamberSlope(x: number, m: number, p: number): number {
  if (m === 0 || p === 0) return 0;
  if (x < p) {
    return ((2 * m) / p ** 2) * (p - x);
  }
  return ((2 * m) / (1 - p) ** 2) * (p - x);
}

function cosineSpacing(nPoints: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < nPoints; i++) {
    const beta = (Math.PI * i) / (nPoints - 1);
    out.push(0.5 * (1 - Math.cos(beta)));
  }
  return out;
}

/** Upper and lower surface coordinates, leading edge to trailing edge, cosine-spaced. */
export function naca4Surface(m: number, p: number, t: number, nPoints = 150): { upper: Point[]; lower: Point[] } {
  const xs = cosineSpacing(nPoints);
  const upper: Point[] = [];
  const lower: Point[] = [];

  for (const x of xs) {
    const yt = nacaThickness(x, t);
    const yc = nacaCamber(x, m, p);
    const dyc = nacaCamberSlope(x, m, p);
    const theta = Math.atan(dyc);

    upper.push({ x: x - yt * Math.sin(theta), y: yc + yt * Math.cos(theta) });
    lower.push({ x: x + yt * Math.sin(theta), y: yc - yt * Math.cos(theta) });
  }

  return { upper, lower };
}

/** Closed boundary loop: lower LE->TE then upper TE->LE (matches naca4_boundary_loop /
 * physics/forces.py's convention). */
export function naca4BoundaryLoop(m: number, p: number, t: number, nPoints = 150): Point[] {
  const { upper, lower } = naca4Surface(m, p, t, nPoints);
  const upperReversed = [...upper].reverse().slice(1);
  return [...lower, ...upperReversed];
}

/** SVG path `d` string for the closed airfoil outline, chord normalized to [0,1]. */
export function naca4SvgPath(m: number, p: number, t: number, nPoints = 100): string {
  const loop = naca4BoundaryLoop(m, p, t, nPoints);
  const [first, ...rest] = loop;
  const cmds = [`M ${first.x.toFixed(5)} ${first.y.toFixed(5)}`];
  for (const pt of rest) {
    cmds.push(`L ${pt.x.toFixed(5)} ${pt.y.toFixed(5)}`);
  }
  cmds.push("Z");
  return cmds.join(" ");
}

export function nacaDesignation(m: number, p: number, t: number): string {
  const d1 = Math.round(m * 100);
  const d2 = Math.round(p * 10);
  const d34 = Math.round(t * 100);
  return `NACA ${d1}${d2}${String(d34).padStart(2, "0")}`;
}
